spec/fixtures/rails6_users_app/appmap.yml

```diff
@@ -5,6 +5,10 @@ packages:
   - path: app/controllers
     labels: [ mvc.controller ]
   - gem: sequel
+functions:
+  - name: logger/Logger::LogDevice#write
+    builtin: true
```

Comment:
I don't entirely understand what makes this function a built-in and why it needs special handling. It's in the standard library, but it's not a native function (at least in my version of the MRI).

Reply:
There has pretty much always been a distinction in appmap-ruby between "builtins" and code that is hooked as it's loaded.

The primary way that appmap-ruby instruments code is by hooking into the TracePoint events that fire as source files are loaded (in response to the `require` chain). But, there is some code that's already loaded before appmap-ruby is initialized. There will never be any TracePoint events for this code. So, the only way to access it (that I know of) is to `require` it and then enumerate the methods on the classes/modules.

To say the same thing a bit differently: We instruct the developer to place the `appmap` gem at the top of the Gemfile, before any other gem. This ensures that all subsequent gems will be observed by the appmap TracePoint handler. But any code that's *not* in a gem - in other words, it's built into Ruby - may already be loaded before appmap is loaded - therefore we can't rely on TracePoint events, and we need to require-then-enumerate that code.

Most of the time, the user is only interested in hooking their own code, and gems. But sometimes, it's useful to hook a builtin - like openssl or logging.

Reply:
In other words: if the class might conceivably be already loaded the user has to use `builtin`? This is a bit confusing, because it's not what builtin means in ruby world and the line isn't always clear. Why should the user know or care about this?

Can we instead simply check for the existence of the target? If it exists, then handle it as "builtin", if not, then add a hook pending future load. (I'll also note that `builtin` path always works, even for code which is in gems.)

Reply:
I get it, but I don't have a great alternative. Because I don't want to go `require`-ing every gem that the user lists in the appmap.yml - because it's entirely possible that a gem has `require: false`, and I am going to screw up the app if I `require` that gem.

Reply:
Why would you need to require it? Either it's already been loaded, and then go ahead and hook is as `builtin`, otherwise add a pending hook that'll fire whenever the module is required in the normal flow. I don't understand why we'd ever need to explicitly load anything here.